gordon_gcp/schema/parse.py

```diff
@@ -0,0 +1,89 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2018 Spotify AB
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Module to parse loaded JSON messages according to GCP-related schemas
+and return only the actionable data.
+
+To use:
+
+.. code-block:: pycon
+
+    >>> import json
+    >>> from gordon_gcp.schema import parse
+    >>> parser = parse.MessageParser()
+    >>> # using an example file
+    >>> exp = 'gordon_gcp/schema/examples/audit-log.last-operation.json'
+    >>> with open(exp, 'r') as r:
+    ...   data = json.load(f)
+    >>> message = parser.parse(data, 'audit-log')
+    >>> message
+    {
+        'action': 'additions',
+        'resourceName': 'projects/.../instances/an-instance-name-b45c',
+        'resourceRecords': []
+    }
+
+"""
+
+
+__all__ = ('MessageParser',)
+
+
+class MessageParser:
+    """Parse a message provided a given GCP schema."""
+    ACTION_MAPPER = {
+        'v1.compute.instances.insert': 'additions',
+        'v1.compute.instances.delete': 'deletions',
+    }
+
+    def _parse_audit_log_msg(self, message):
+        # TODO (lynn): what about first vs last messages?
+        payload = message['protoPayload']
+        data = {
+            'action': self.ACTION_MAPPER[payload['methodName']],
+            'resourceName': payload['resourceName'],
+            'resourceRecords': [],
+        }
+        return data
+
+    def _parse_event_msg(self, message):
+        # TODO (lynn): I think this needs to be a list
+        records = []
+        record = message['resourceRecords']
+        records.append(record)
+        return {
+            'action': message['action'],
+            'resourceName': None,
+            'resourceRecords': [message['resourceRecords']]
+        }
+
+    def parse(self, message, schema):
+        """Parse message according to schema.
+
+        `message` should already be validated against the given schema.
+        See :ref:`schemadef` for more information.
+
+        Args:
+            message (dict): message data to parse.
+            schema (str): valid message schema.
+        Returns:
+            (dict): parsed message
+        """
+        func = {
+            'audit-log': self._parse_audit_log_msg,
+            'event': self._parse_event_msg,
+        }[schema]
```

Comment:
Could this map be a object var like the action mapper (I'm unclear on the role of `self` here)?

Reply:
It needs self to have access to the methods that this class defines.

Reply:
These seem to be essentially static functions. What is the advantage of writing them as instance methods?

Reply:
Only `_parse_event_msg` could be static, since `_parse_audit_log_msg` needs access to `self`. Though it could be a `classmethod` since it just needs access to the `ACTION_MAPPER`.

I'm not sure of the benefit for static/class methods versus instance methods. They are private methods, so only `parse` should be calling them. But maybe I'm missing something that you're thinking of?

Reply:
I was just thinking that it might be more complicated than it needs to be.  Why instantiate an object and pass around self, if it isn't needed? Or even why attach these functions to a object/class at all, rather than having `parse` as a module function?

Reply:
The key driver for me across the board is to make everything extensible. This way, we don't restrict how users parse a message, i.e. maybe users want to access tags that are available in the audit log, or add a schema to parse against. I might be over-optimizing. @matt0bi any thoughts on the matter?